File: WinPaletter/Extras/AltTabEditor.vb
Imports System.ComponentModel
Imports WinPaletter.PreviewHelpers

Public Class AltTabEditor
    Private Sub AltTabEditor_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        LoadLanguage
        ApplyStyle(Me)
        Button12.Image = MainFrm.Button20.Image.Resize(16, 16)
        ApplyFromCP(My.CP)

        Select Case My.PreviewStyle
            Case WindowStyle.W11
                RadioImage1.Image = My.Resources.Native11

            Case WindowStyle.W10
                RadioImage1.Image = My.Resources.Native10

            Case WindowStyle.W81
                RadioImage1.Image = My.Resources.Native8

            Case WindowStyle.W7
                RadioImage1.Image = My.Resources.Native7

            Case WindowStyle.WVista
                RadioImage1.Image = My.Resources.NativeVista

            Case Else
                RadioImage1.Image = My.Resources.Native11

        End Select

        RadioImage2.Image = My.Resources.NativeXP

        pnl_preview1.BackgroundImage = MainFrm.pnl_preview.BackgroundImage
        Classic_Preview1.BackgroundImage = MainFrm.pnl_preview_classic.BackgroundImage

        SetClassicPanelRaisedRColors(My.CP, PanelRRaised1)
        SetClassicPanelColors(My.CP, PanelR1)

        Panel1.BackColor = My.CP.Win32.Hilight

        Select Case My.PreviewStyle
            Case WindowStyle.W11
                WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab11
                WinElement1.DarkMode = Not My.CP.Windows11.WinMode_Light

            Case WindowStyle.W10
                WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10
                WinElement1.DarkMode = Not My.CP.Windows10.WinMode_Light

            Case WindowStyle.W81
                Select Case My.CP.Windows81.Theme
                    Case CP.Structures.Windows7.Themes.Aero
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab8Aero
                        WinElement1.BackColor = My.CP.Windows81.PersonalColors_Background
                        WinElement1.Background2 = My.CP.Windows81.PersonalColors_Background

                    Case CP.Structures.Windows7.Themes.AeroLite
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab8AeroLite
                        WinElement1.BackColor = My.CP.Win32.Window
                        WinElement1.Background2 = My.CP.Win32.Hilight
                        WinElement1.LinkColor = My.CP.Win32.ButtonText
                        WinElement1.ForeColor = My.CP.Win32.WindowText

                End Select

            Case WindowStyle.W7
                Select Case My.CP.Windows7.Theme
                    Case CP.Structures.Windows7.Themes.Aero
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Aero

                    Case CP.Structures.Windows7.Themes.AeroOpaque
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Opaque

                    Case CP.Structures.Windows7.Themes.Basic
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Basic

                End Select

                WinElement1.BackColor = My.CP.Windows7.ColorizationColor
                WinElement1.Background2 = My.CP.Windows7.ColorizationAfterglow
                WinElement1.BackColorAlpha = My.CP.Windows7.ColorizationBlurBalance
                WinElement1.Win7ColorBal = My.CP.Windows7.ColorizationColorBalance
                WinElement1.Win7GlowBal = My.CP.Windows7.ColorizationAfterglowBalance
                WinElement1.NoisePower = My.CP.Windows7.ColorizationGlassReflectionIntensity
                WinElement1.Shadow = My.CP.WindowsEffects.WindowShadow
        End Select

        Panel2.BackColor = PanelRRaised1.BackColor
        LabelR1.Font = My.CP.MetricsFonts.CaptionFont

        GroupBox4.Enabled = (WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10) Or ExplorerPatcher.IsAllowed
        AlertBox1.Visible = (My.PreviewStyle = WindowStyle.W7)

        If ExplorerPatcher.IsAllowed Then
            Try
                If My.Computer.Registry.CurrentUser.GetValue("Software\Microsoft\Windows\CurrentVersion\Explorer\AltTabSettings", 0) = 3 Then
                    WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10
                    WinElement1.DarkMode = Not My.CP.Windows11.WinMode_Light
                End If
            Finally
                My.Computer.Registry.CurrentUser.Close()
            End Try
        End If

        If WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Basic Then
            WinElement1.Size = New Size(360, 100)
        Else
            WinElement1.Size = New Size(450, 150)
        End If

        WinElement1.Left = (WinElement1.Parent.Width - WinElement1.Width) / 2
        WinElement1.Top = (WinElement1.Parent.Height - WinElement1.Height) / 2

        tabs_preview_1.DoubleBuffer
    End Sub

    Sub ApplyFromCP(CP As CP)
        With CP.AltTab
            AltTabEnabled.Checked = .Enabled
            RadioImage1.Checked = .Style = CP.Structures.AltTab.Styles.Default Or .Style = CP.Structures.AltTab.Styles.EP_Win10
            RadioImage2.Checked = .Style = CP.Structures.AltTab.Styles.ClassicNT
            Trackbar1.Value = .Win10Opacity
        End With
    End Sub

    Sub ApplyToCP(CP As CP)
        With CP.AltTab
            .Enabled = AltTabEnabled.Checked
            .Style = If(RadioImage1.Checked, CP.Structures.AltTab.Styles.Default, CP.Structures.AltTab.Styles.ClassicNT)
            If ExplorerPatcher.IsAllowed And WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10 Then .Style = CP.Structures.AltTab.Styles.EP_Win10
            .Win10Opacity = Trackbar1.Value
        End With
    End Sub

    Private Sub Button11_Click(sender As Object, e As EventArgs) Handles Button11.Click
        If OpenFileDialog1.ShowDialog = DialogResult.OK Then
            Dim CPx As New CP(CP.CP_Type.File, OpenFileDialog1.FileName)
            ApplyFromCP(CPx)
            CPx.Dispose()
        End If
    End Sub

    Private Sub Button9_Click(sender As Object, e As EventArgs) Handles Button9.Click
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyFromCP(CPx)
        CPx.Dispose()
    End Sub

    Private Sub Button12_Click(sender As Object, e As EventArgs) Handles Button12.Click
        Using _Def As CP = CP_Defaults.From(My.PreviewStyle)
            ApplyFromCP(_Def)
        End Using
    End Sub

    Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click
        Me.Close()
    End Sub

    Private Sub Button10_Click(sender As Object, e As EventArgs) Handles Button10.Click
        Cursor = Cursors.WaitCursor
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyToCP(CPx)
        ApplyToCP(My.CP)
        CPx.AltTab.Apply()
        CPx.Dispose()
        Cursor = Cursors.Default
    End Sub

    Private Sub Button8_Click(sender As Object, e As EventArgs) Handles Button8.Click
        ApplyToCP(My.CP)
        Me.Close()
    End Sub

    Private Sub AltTabEnabled_CheckedChanged(sender As Object, e As EventArgs) Handles AltTabEnabled.CheckedChanged
        checker_img.Image = If(sender.Checked, My.Resources.checker_enabled, My.Resources.checker_disabled)
    End Sub

    Private Sub Opacity_btn_Click(sender As Object, e As EventArgs) Handles opacity_btn.Click
        Dim response As String = InputBox(My.Lang.InputValue, sender.text, My.Lang.ItMustBeNumerical)
        sender.Text = Math.Max(Math.Min(Val(response), Trackbar1.Maximum), Trackbar1.Minimum) : Trackbar1.Value = Val(sender.Text)
    End Sub

    Private Sub Trackbar1_Scroll(sender As Object) Handles Trackbar1.Scroll
        opacity_btn.Text = sender.Value.ToString
        If WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10 Then WinElement1.BackColorAlpha = Trackbar1.Value
    End Sub

    Private Sub RadioImage2_CheckedChanged(sender As Object) Handles RadioImage2.CheckedChanged
        If RadioImage2.Checked Then tabs_preview_1.SelectedIndex = 1
    End Sub

    Private Sub RadioImage1_CheckedChanged(sender As Object) Handles RadioImage1.CheckedChanged
        If RadioImage1.Checked Then tabs_preview_1.SelectedIndex = 0
    End Sub

    Private Sub Form_HelpButtonClicked(sender As Object, e As CancelEventArgs) Handles Me.HelpButtonClicked
        Process.Start(My.Resources.Link_Wiki & "/Edit-Windows-switcher-(Alt-Tab-appearance)")
    End Sub
End Class

File: WinPaletter/Extras/AltTabEditor.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Windows.Forms;
using static WinPaletter.PreviewHelpers;

namespace WinPaletter
{
    public partial class AltTabEditor
    {
        private void Form_HelpButtonClicked(object sender, CancelEventArgs e)
        {
            Process.Start($"{Properties.Resources.Link_Wiki}/Edit-Windows-switcher-(Alt-Tab-appearance)");
        }

        public AltTabEditor()
        {
            InitializeComponent();
        }

        private void LoadFromWPTH(object sender, EventArgs e)
        {
            if (OpenFileDialog1.ShowDialog() == DialogResult.OK)
            {
                Theme.Manager TMx = new(Theme.Manager.Source.File, OpenFileDialog1.FileName);
                LoadFromTM(TMx);
                TMx.Dispose();
            }
        }

        private void LoadFromCurrent(object sender, EventArgs e)
        {
            Theme.Manager TMx = new(Theme.Manager.Source.Registry);
            LoadFromTM(TMx);
            TMx.Dispose();
        }

        private void LoadFromDefault(object sender, EventArgs e)
        {
            Theme.Manager TMx = Theme.Default.Get(Program.WindowStyle);
            LoadFromTM(TMx);
            TMx.Dispose();
        }

        private void LoadIntoCurrentTheme(object sender, EventArgs e)
        {
            ApplyToTM(Program.TM);
            Close();
        }

        private void Apply(object sender, EventArgs e)
        {
            Cursor = Cursors.WaitCursor;

            using (Theme.Manager TMx = new(Theme.Manager.Source.Registry))
            {
                ApplyToTM(TMx);
                ApplyToTM(Program.TM);
                TMx.AltTab.Apply();
            }

            Cursor = Cursors.Default;
        }

        private void AltTabEditor_Load(object sender, EventArgs e)
        {
            DesignerData data = new(this)
            {
                AspectName = Program.Lang.Store_Toggle_AltTab,
                Enabled = Program.TM.AltTab.Enabled,
                Import_theme = false,
                Import_msstyles = false,
                GeneratePalette = false,
                GenerateMSTheme = false,
                Import_preset = false,
                CanSwitchMode = false,

                OnLoadIntoCurrentTheme = LoadIntoCurrentTheme,
                OnApply = Apply,
                OnImportFromDefault = LoadFromDefault,
                OnImportFromWPTH = LoadFromWPTH,
                OnImportFromCurrentApplied = LoadFromCurrent,
            };

            LoadData(data);

            LoadFromTM(Program.TM);

            switch (Program.WindowStyle)
            {
                case WindowStyle.W12:
                    {
                        RadioImage1.Image = Assets.WinLogos.Win12;
                        break;
                    }

                case WindowStyle.W11:
                    {
                        RadioImage1.Image = Assets.WinLogos.Win11;
                        break;
                    }

                case WindowStyle.W10:
                    {
                        RadioImage1.Image = Assets.WinLogos.Win10;
                        break;
                    }

                case WindowStyle.W81:
                    {
                        RadioImage1.Image = Assets.WinLogos.Win81;
                        break;
                    }

                case WindowStyle.W7:
                    {
                        RadioImage1.Image = Assets.WinLogos.Win7;
                        break;
                    }

                case WindowStyle.WVista:
                    {
                        RadioImage1.Image = Assets.WinLogos.WinVista;
                        break;
                    }

                case WindowStyle.WXP:
                    {
                        RadioImage1.Image = Assets.WinLogos.WinXP;
                        break;
                    }

                default:
                    {
                        RadioImage1.Image = Assets.WinLogos.Win12;
                        break;
                    }
            }

            RadioImage2.Image = Assets.WinLogos.WinXP;

            pnl_preview1.BackgroundImage = Program.Wallpaper;
            Classic_Preview1.BackgroundImage = Program.Wallpaper;

            SetClassicPanelRaisedRColors(Program.TM, PanelRRaised1);
            SetClassicPanelColors(Program.TM, PanelR1);

            Panel1.BackColor = Program.TM.Win32.Hilight;

            switch (Program.WindowStyle)
            {
                case WindowStyle.W12:
                    {
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab11;
                        WinElement1.DarkMode = !Program.TM.Windows12.WinMode_Light;
                        break;
                    }

                case WindowStyle.W11:
                    {
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab11;
                        WinElement1.DarkMode = !Program.TM.Windows11.WinMode_Light;
                        break;
                    }

                case WindowStyle.W10:
                    {
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10;
                        WinElement1.DarkMode = !Program.TM.Windows10.WinMode_Light;
                        break;
                    }

                case WindowStyle.W81:
                    {
                        switch (Program.TM.Windows81.Theme)
                        {
                            case Theme.Structures.Windows7.Themes.Aero:
                                {
                                    WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab8Aero;
                                    WinElement1.BackColor = Program.TM.Windows81.PersonalColors_Background;
                                    WinElement1.Background2 = Program.TM.Windows81.PersonalColors_Background;
                                    break;
                                }

                            case Theme.Structures.Windows7.Themes.AeroLite:
                                {
                                    WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab8AeroLite;
                                    WinElement1.BackColor = Program.TM.Win32.Window;
                                    WinElement1.Background2 = Program.TM.Win32.Hilight;
                                    WinElement1.LinkColor = Program.TM.Win32.ButtonText;
                                    WinElement1.ForeColor = Program.TM.Win32.WindowText;
                                    break;
                                }

                        }

                        break;
                    }

                case WindowStyle.W7:
                    {
                        switch (Program.TM.Windows7.Theme)
                        {
                            case Theme.Structures.Windows7.Themes.Aero:
                                {
                                    WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Aero;
                                    break;
                                }

                            case Theme.Structures.Windows7.Themes.AeroOpaque:
                                {
                                    WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Opaque;
                                    break;
                                }

                            case Theme.Structures.Windows7.Themes.Basic:
                                {
                                    WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab7Basic;
                                    break;
                                }

                        }

                        WinElement1.BackColor = Program.TM.Windows7.ColorizationColor;
                        WinElement1.Background2 = Program.TM.Windows7.ColorizationAfterglow;
                        WinElement1.BackColorAlpha = Program.TM.Windows7.ColorizationBlurBalance;
                        WinElement1.Win7ColorBal = Program.TM.Windows7.ColorizationColorBalance;
                        WinElement1.Win7GlowBal = Program.TM.Windows7.ColorizationAfterglowBalance;
                        WinElement1.NoisePower = Program.TM.Windows7.ColorizationGlassReflectionIntensity;
                        WinElement1.Shadow = Program.TM.WindowsEffects.WindowShadow;
                        break;
                    }
            }

            Panel2.BackColor = PanelRRaised1.BackColor;
            LabelR1.Font = Program.TM.MetricsFonts.CaptionFont;

            GroupBox4.Enabled = WinElement1.Style == UI.Simulation.WinElement.Styles.AltTab10 || ExplorerPatcher.IsAllowed();
            AlertBox1.Visible = Program.WindowStyle == WindowStyle.W7;

            if (ExplorerPatcher.IsAllowed())
            {
                try
                {
                    if (Convert.ToInt32(GetReg(@"HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer", "AltTabSettings", 0)) == 3)
                    {
                        WinElement1.Style = UI.Simulation.WinElement.Styles.AltTab10;
                        WinElement1.DarkMode = !Program.TM.Windows11.WinMode_Light;
                    }
                }
                finally
                {
                    Microsoft.Win32.Registry.CurrentUser.Close();
                }
            }

            if (WinElement1.Style == UI.Simulation.WinElement.Styles.AltTab7Basic)
            {
                WinElement1.Size = new(360, 100);
            }
            else
            {
                WinElement1.Size = new(450, 150);
            }

            WinElement1.Left = (int)Math.Round((WinElement1.Parent.Width - WinElement1.Width) / 2d);
            WinElement1.Top = (int)Math.Round((WinElement1.Parent.Height - WinElement1.Height) / 2d);

            tabs_preview_1.DoubleBuffer();
        }

        public void LoadFromTM(Theme.Manager TM)
        {
            AspectEnabled = TM.AltTab.Enabled;
            RadioImage1.Checked = TM.AltTab.Style == Theme.Structures.AltTab.Styles.Default | TM.AltTab.Style == Theme.Structures.AltTab.Styles.EP_Win10;
            RadioImage2.Checked = TM.AltTab.Style == Theme.Structures.AltTab.Styles.ClassicNT;
            Trackbar1.Value = TM.AltTab.Win10Opacity;
        }

        public void ApplyToTM(Theme.Manager TM)
        {
            TM.AltTab.Enabled = AspectEnabled;
            TM.AltTab.Style = RadioImage1.Checked ? Theme.Structures.AltTab.Styles.Default : Theme.Structures.AltTab.Styles.ClassicNT;
            if (ExplorerPatcher.IsAllowed() & WinElement1.Style == UI.Simulation.WinElement.Styles.AltTab10)
                TM.AltTab.Style = Theme.Structures.AltTab.Styles.EP_Win10;
            TM.AltTab.Win10Opacity = Trackbar1.Value;
        }

        private void RadioImage2_CheckedChanged(object sender, EventArgs e)
        {
            if (RadioImage2.Checked)
                tabs_preview_1.SelectedIndex = 1;
        }

        private void RadioImage1_CheckedChanged(object sender, EventArgs e)
        {
            if (RadioImage1.Checked)
                tabs_preview_1.SelectedIndex = 0;
        }

        private void trackBarX1_ValueChanged(object sender, EventArgs e)
        {
            if (WinElement1.Style == UI.Simulation.WinElement.Styles.AltTab10) { WinElement1.BackColorAlpha = Trackbar1.Value; }
        }
    }
}